framework/src/userobject/RenormalizeVector.C

```diff
@@ -0,0 +1,195 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "RenormalizeVector.h"
+#include "MooseError.h"
+#include "NonlinearSystemBase.h"
+
+#include "libmesh/numeric_vector.h"
+#include "libmesh/int_range.h"
+
+registerMooseObject("MooseApp", RenormalizeVector);
+
+class RenormalizeVectorThread : public ParallelObject
+{
+public:
+  RenormalizeVectorThread(FEProblemBase & problem,
+                          const std::vector<VariableName> & vars,
+                          const Real norm);
+  // Splitting Constructor
+  RenormalizeVectorThread(RenormalizeVectorThread & x, Threads::split split);
+  // destructor to close the solution once more
+  ~RenormalizeVectorThread();
+
+  void operator()(const ConstElemRange & range);
+  void join(const RenormalizeVectorThread &) {}
+
+protected:
+  FEProblemBase & _problem;
+  System * _sys;
+  std::vector<unsigned int> _var_numbers;
+  const Real _target_norm;
+
+  THREAD_ID _tid;
+};
+
+InputParameters
+RenormalizeVector::validParams()
+{
+  InputParameters params = GeneralUserObject::validParams();
+  params.addClassDescription("Renormalize the solution of a set of variables comprising a vector");
+  params.addCoupledVar("v", "Variables comprising the vector");
+  params.addParam<Real>("norm", 1.0, "Desired norm for the coupled variable vector");
+  return params;
+}
+
+RenormalizeVector::RenormalizeVector(const InputParameters & parameters)
+  : GeneralUserObject(parameters),
+    _mesh(_fe_problem.mesh()),
+    _vars(getParam<std::vector<VariableName>>("v")),
+    _target_norm(getParam<Real>("norm"))
+{
```

Comment:
should we error if there s a single variable? it s not the most exciting object for single variable

Reply:
No, because passing in a single array variable is permitted and useful. And passing in a single scalar is not going to give you a wrong result either. You don't know that this wouldn't be useful. For example reducing a 2d simulation to a 1d simulation could be very useful during development.